alg/teca_temporal_reduction.py

```diff
@@ -0,0 +1,506 @@
+import sys
+import teca_py
+import numpy as np
+
+
+class time_point:
+    """
+    A structure holding a floating point time value and its
+    corresponding year, month day, hour minute and second
+    """
+    def __init__(self, t, units, calendar):
+        self.t = t
+        self.units = units
+        self.calendar = calendar
+
+        self.year, self.month, self.day, \
+            self.hour, self.minutes, self.seconds = \
+            teca_py.calendar_util.date(t, self.units, self.calendar)
+
+    def __str__(self):
+        return '%g (%s, %s) --> %04d-%02d-%02d %02d:%02d:%02g' % (
+            self.t, self.units, self.calendar, self.year, self.month, self.day,
+            self.hour, self.minutes, self.seconds)
+
+
+class c_struct:
+    """
+    A c like data structure
+    """
+    def __init__(self, **kwds):
+        self.__dict__.update(kwds)
+
+    def __str__(self):
+        strg = ''
+        for k, v in self.__dict__.items():
+            strg += k + '=' + str(v) + ', '
+        return strg
+
+
+class interval_iterator:
+    class month_iterator:
+        """
+        An iterator over all months between 2 time_point's. A pair
+        of time steps bracketing the current month are returned at
+        each iteration.
+        """
+
+        def __init__(self, t, units, calendar):
+            """
+            t - an array of floating point time values
+            units - string units of the time values
+            calendar - string name of the calendar system
+            """
+            self.t = t
+            self.units = units
+
+            calendar = calendar.lower()
+            self.calendar = calendar
+
+            self.noleap = \
+                True if (calendar == 'noleap') or (calendar == 'no_leap') \
+                or (calendar == '365_day') or (calendar == '360_day') else \
+                False
+
+            # time point's to iterate between
+            self.t0 = time_point(t[0], units, calendar)
+            self.t1 = time_point(t[-1], units, calendar)
+
+            # current time state
+            self.year = self.t0.year
+            self.month = self.t0.month
+
+        def last_day_of_month(self):
```

Comment:
can we use the `calcalcs` `ccs_dpm()` routine in here instead?  Examining `calcalcs.cxx` shows that the logic for computing this is substantially more complicated than just considering leap years.  For example, the 360_day calendar should have 30 days in all months.

Reply:
I exposed this functionality in a new function `calcalcs::days_in_month`